File: ios/versioned/sdk49/ExpoMaps/ExpoMaps/GoogleMaps/GoogleMapsCameraAnimations.swift
import Foundation
import ABI49_0_0ExpoModulesCore
import GoogleMaps

class GoogleMapsCameraAnimations {
  private let mapView: GMSMapView

  init(mapView: GMSMapView) {
    self.mapView = mapView
  }

  func moveCamera(cameraMove: CameraMoveRecord, promise: Promise?) {
    var boundsUpdate: GMSCameraUpdate?
    let target = CLLocationCoordinate2D(
      latitude: cameraMove.target["latitude"] as? CLLocationDegrees ?? mapView.camera.target.latitude,
      longitude: cameraMove.target["longitude"] as? CLLocationDegrees ?? mapView.camera.target.longitude
    )

    if let delta = cameraMove.latLngDelta {
      let latitudeDelta = delta.latitudeDelta
      let longitudeDelta = delta.longitudeDelta
      let x1 = target.latitude - latitudeDelta / 2
      let y1 = target.longitude - longitudeDelta / 2
      let topLeft = CLLocationCoordinate2D(latitude: x1 + latitudeDelta, longitude: y1)
      let bottomRight = CLLocationCoordinate2D(latitude: x1, longitude: y1 + longitudeDelta)
      boundsUpdate = GMSCameraUpdate.fit(GMSCoordinateBounds(coordinate: topLeft, coordinate: bottomRight))
    }

    if cameraMove.animate {
      CATransaction.begin()
      CATransaction.setCompletionBlock {
        /*
         Calculating the projection while next animation is starting sometimes causes the map view to flicker,
          the projection isn't accessed in order to avoid this
        */
        promise?.resolve(CameraPositionRecord(cameraPosition: self.mapView.camera, visibleRegion: nil).toDictionary())
      }
      CATransaction.setAnimationDuration(Double(cameraMove.duration) / 1000)
      CATransaction.setAnimationTimingFunction(CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeInEaseOut))

      if let tilt = cameraMove.tilt {
        mapView.animate(toViewingAngle: tilt)
      }
      if let bearing = cameraMove.bearing {
        mapView.animate(toBearing: bearing)
      }
      if let zoom = cameraMove.zoom {
        mapView.animate(toZoom: zoom)
      }
      mapView.animate(toLocation: target)

      // if bounds are set this will override all of the properties set above, except for the target
      if let boundsUpdate = boundsUpdate {
        mapView.animate(with: boundsUpdate)
      }

      CATransaction.commit()
    } else {
      if let boundsUpdate = boundsUpdate {
        mapView.moveCamera(boundsUpdate)
      } else {
        mapView.camera = GMSCameraPosition(
          latitude: target.latitude,
          longitude: target.longitude,
          zoom: cameraMove.zoom ?? mapView.camera.zoom,
          bearing: cameraMove.bearing ?? mapView.camera.bearing,
          viewingAngle: cameraMove.tilt ?? mapView.camera.viewingAngle
        )
      }
      promise?.resolve(CameraPositionRecord(cameraPosition: mapView.camera, visibleRegion: mapView.projection.visibleRegion()).toDictionary())
    }
  }
}

File: packages/expo-maps/android/src/main/java/expo/modules/maps/googleMaps/GoogleMapsCameraAnimations.kt
package expo.modules.maps.googleMaps

import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import expo.modules.kotlin.Promise
import expo.modules.maps.records.CameraMoveRecord
import expo.modules.maps.records.CameraPositionRecord

class GoogleMapsCameraAnimations(private val googleMap: GoogleMap) {
  fun moveCamera(cameraMove: CameraMoveRecord, promise: Promise?) {
    val cameraPositionBuilder = CameraPosition.Builder()
    var boundsUpdate: CameraUpdate? = null
    val target = LatLng(
      if (cameraMove.target?.latitude != null) cameraMove.target!!.latitude else googleMap.cameraPosition.target.latitude,
      if (cameraMove.target?.longitude != null) cameraMove.target!!.longitude else googleMap.cameraPosition.target.longitude
    )

    if (cameraMove.latLngDelta != null) {
      val x1 = target.latitude - cameraMove.latLngDelta!!.latitudeDelta / 2
      val y1 = target.longitude - cameraMove.latLngDelta!!.longitudeDelta / 2
      val bottomLeft = LatLng(x1, y1)
      val topRight = LatLng(
        x1 + cameraMove.latLngDelta!!.latitudeDelta,
        y1 + cameraMove.latLngDelta!!.longitudeDelta
      )
      boundsUpdate = CameraUpdateFactory.newLatLngBounds(LatLngBounds(bottomLeft, topRight), 0)
    }

    cameraPositionBuilder.target(target)

    if (cameraMove.zoom != null) {
      cameraPositionBuilder.zoom(cameraMove.zoom!!)
    }
    if (cameraMove.bearing != null) {
      cameraPositionBuilder.bearing(cameraMove.bearing!!)
    }
    if (cameraMove.tilt != null) {
      cameraPositionBuilder.tilt(cameraMove.tilt!!)
    }
    val cameraPosition = cameraPositionBuilder.build()

    val resolvePromise = object : GoogleMap.CancelableCallback {
      override fun onCancel() {
        promise?.resolve(CameraPositionRecord(googleMap.cameraPosition))
      }

      override fun onFinish() {
        promise?.resolve(CameraPositionRecord(googleMap.cameraPosition))
      }
    }
    if (cameraMove.animate) {
      if (boundsUpdate == null) {
        googleMap.animateCamera(
          CameraUpdateFactory.newCameraPosition(cameraPosition),
          cameraMove.duration,
          resolvePromise
        )
      } else {
        googleMap.animateCamera(
          boundsUpdate,
          cameraMove.duration,
          resolvePromise
        )
      }
    } else {
      if (boundsUpdate == null) {
        googleMap.moveCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
      } else {
        googleMap.moveCamera(boundsUpdate)
      }
      promise?.resolve(CameraPositionRecord(googleMap.cameraPosition))
    }
  }
}
